### PR TITLE
Unify sbt shell and build.sbt syntax (scope path syntax)

### PR DESCRIPTION
**Pull request by eed3si9n**
Fixes sbt/sbt#1812

This adds scope path syntax (fka unified slash syntax) for both sbt shell and the build.sbt DSL.

### Scope path syntax in shell

Instead of the current `<project-id>/config:intask::key`, this **adds** `<project-id>/<config-ident>/intask/key` where `<config-ident>` is the Scala identifier notation for the configurations like `Compile` and `Test`. (The old shell syntax will continue to function)

### Scope path syntax in build.sbt

This also adds a series of implicits called `SlashSyntax` that adds `/` operators to project references, configuration, and keys such that the same syntax works in build.sbt.

### Examples

These examples work *both* from the shell *and* in build.sbt.

```scala
Global / cancelable
ThisBuild / scalaVersion
Test / test
root / Compile / compile / scalacOptions
ProjectRef(uri("file:/xxx/helloworld/"),"root")/Compile/scalacOptions
Zero / Zero / name
```

The inspect command now outputs something that can be copy-pasted:

    > inspect compile
    [info] Task: sbt.inc.Analysis
    [info] Description:
    [info] 	Compiles sources.
    [info] Provided by:
    [info] 	ProjectRef(uri("file:/xxx/helloworld/"), "root") / Compile / compile
    [info] Defined at:
    [info] 	(sbt.Defaults) Defaults.scala:326
    [info] Dependencies:
    [info] 	Compile / manipulateBytecode
    [info] 	Compile / incCompileSetup
    [info] Reverse dependencies:
    [info] 	Compile / printWarnings
    [info] 	Compile / products
    [info] 	Compile / discoveredSbtPlugins
    [info] 	Compile / discoveredMainClasses
    [info] Delegates:
    [info] 	Compile / compile
    [info] 	compile
    [info] 	ThisBuild / Compile / compile
    [info] 	ThisBuild / compile
    [info] 	Zero / Compile / compile
    [info] 	Global / compile
    [info] Related:
    [info] 	Test / compile

slash

### CODE REVIEW

**Comment by jvican** [comment]
Two questions:
  
1. What's missing on this front for this to be merged?
2. Is there any chance this can make it into the next sbt release?

**Comment by eed3si9n** [comment]
1. I need to add mima exclusion, I think to pass Travis.
2. I am hoping it will be in sbt 1.1.0, 1.2.0 or some feature release. Immediately next one will be sbt 1.0.2 (hotfix release).

**Comment by dwijnand** [comment]
I've added this to the 1.1.0 milestone as it would be a good highlight feature of sbt 1.1.

**Comment by typesafe-tools** [comment]
To validate this pull request against other sbt modules, please visit <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild/?sbt=pull/3434/head&deploy_to_bintray=false&run_sbt_community=false">this link.</A>

**Comment by typesafe-tools** [comment]
To validate this pull request against other sbt modules, please visit <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild/?sbt=pull/3434/head&deploy_to_bintray=false&run_sbt_community=false">this link.</A>

**Comment by eed3si9n** [comment]
Rebased to the latest 1.x branch, and added Mima exclusion.

**Comment by typesafe-tools** [comment]
To validate this pull request against other sbt modules, please visit <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild/?sbt=pull/3434/head&deploy_to_bintray=false&run_sbt_community=false">this link.</A>

**Comment by typesafe-tools** [comment]
To validate this pull request against other sbt modules, please visit <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild/?sbt=pull/3434/head&deploy_to_bintray=false&run_sbt_community=false">this link.</A>

**Comment by typesafe-tools** [comment]
To validate this pull request against other sbt modules, please visit <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild/?sbt=pull/3434/head&deploy_to_bintray=false&run_sbt_community=false">this link.</A>

**Comment by dwijnand** [comment]
Could you squash the commits?

File: sbt/src/sbt-test/project/unified/test [hunk 1]
```diff
@@ -0,0 +1,13 @@
+> root/Compile/compile
+
+# "Global" now works in shell. optional whitespace arond /
+> show Global / cancelable
+
+# "ThisBuild" now works in shell. optional whitespace arond /
+> show ThisBuild / scalaVersion
+
+> show Test / test
+
```
Comment (jvican) [inline sbt/src/sbt-test/project/unified/test]: `inspect`, `inspect-tree` and other commands taking keys should be tested.

File: sbt/src/sbt-test/project/unified/test [hunk 2]
```diff
@@ -0,0 +1,13 @@
+> root/Compile/compile
+
+# "Global" now works in shell. optional whitespace arond /
+> show Global / cancelable
+
+# "ThisBuild" now works in shell. optional whitespace arond /
```
Comment (jvican) [inline sbt/src/sbt-test/project/unified/test]: The optional whitespace around `/` should be tested for every example.
  
Also, typo in "arond".

File: sbt/src/sbt-test/project/unified/test [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+> show ThisBuild / scalaVersion
+
+> show Test / test
```
Comment (jvican) [inline sbt/src/sbt-test/project/unified/test]: How do we know that we're calling the right `Test / test`? This tests that the parser accepts it and it runs something, but we don't know which something has been run.

**Review by jvican — COMMENTED** [review]
Comments in tests.

**Comment by typesafe-tools** [comment]
To validate this pull request against other sbt modules, please visit <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild/?sbt=pull/3434/head&deploy_to_bintray=false&run_sbt_community=false">this link.</A>

**Comment by eed3si9n** [comment]
@jvican @dwijnand Thanks guys for the review.

File: sbt/src/sbt-test/project/unified/build.sbt [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+    ThisBuild / scalaVersion := "2.12.3",
+    console / scalacOptions += "-deprecation",
+    Compile / console / scalacOptions += "-Ywarn-numeric-widen",
+    projA / Compile / console / scalacOptions += "-feature",
+    Zero / Zero / name := "foo",
+
+    libraryDependencies += uTest % Test,
+    testFrameworks += new TestFramework("utest.runner.Framework"),
+
+    commands += Command("inspectCheck", inspectBrief, inspectDetailed)(Inspect.parser) {
```
Comment (eed3si9n) [inline sbt/src/sbt-test/project/unified/build.sbt:17]: @jvican Added the inspect output test here.
Since the parser is shared, I didn't feel the need to test various modes for inspect, but the default mode has displays the keys, so I've added the test.

File: sbt/src/sbt-test/project/unified/test [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+# "ThisBuild" now works in shell.
+> show ThisBuild/scalaVersion
+
+# "ThisBuild" now works in shell. optional whitespace around /
+> show ThisBuild / scalaVersion
+
+$ mkdir target
+> Test/test
+# Check the side-effect of executing uTest
+$ exists target/foo
```
Comment (eed3si9n) [inline sbt/src/sbt-test/project/unified/test:18]: @jvican Added a small "test" that should check that that uTest in fact ran.

File: sbt/src/sbt-test/project/unified/test [hunk 6]
```diff
@@ -0,0 +1,28 @@
+> root/Compile/compile
+
+# "Global" now works in shell.
+> show Global/cancelable
+
+# "Global" now works in shell. optional whitespace around /
+> show Global / cancelable
+
+# "ThisBuild" now works in shell.
+> show ThisBuild/scalaVersion
```
Comment (eed3si9n) [inline sbt/src/sbt-test/project/unified/test:10]: @jvican Added variants with and without whitespace around /.

**Comment by typesafe-tools** [comment]
To validate this pull request against other sbt modules, please visit <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild/?sbt=pull/3434/head&deploy_to_bintray=false&run_sbt_community=false">this link.</A>

File: main-settings/src/main/scala/sbt/Reference.scala [hunk 7]
```diff
@@ -91,7 +91,7 @@ object Reference {
       case LocalRootProject    => "{<this>}<root>"
       case LocalProject(id)    => "{<this>}" + id
       case RootProject(uri)    => "{" + uri + " }<root>"
-      case ProjectRef(uri, id) => "{" + uri + "}" + id
+      case ProjectRef(uri, id) => s"""ProjectRef(uri("$uri"),"$id")"""
```
Comment (dwijnand) [inline main-settings/src/main/scala/sbt/Reference.scala]: could you add a space, please? `s"""ProjectRef(uri("$uri"), "$id")"""`

File: main-settings/src/main/scala/sbt/Def.scala [hunk 8]
```diff
@@ -57,17 +57,36 @@ object Def extends Init[Scope] with TaskMacroExtra {
...
     project match {
-      case BuildRef(current.build)      => "{.}/"
-      case `current`                    => if (multi) current.project + "/" else ""
-      case ProjectRef(current.build, x) => x + "/"
-      case _                            => Reference.display(project) + "/"
+      case BuildRef(current.build) => "ThisBuild" + slash
+      case `current`               =>
+        // ignore (multi) and always display relative key
+        // if (multi) current.project + slash else ""
+        ""
```
Comment (dwijnand) [inline main-settings/src/main/scala/sbt/Def.scala]: Could you explain what's going on here? Why is multi ignored now?

File: main-settings/src/main/scala/sbt/Scope.scala [hunk 9]
```diff
@@ -123,29 +123,84 @@ object Scope {
...
-      mask: ScopeMask
-  ): String = {
+  def displayMasked(scope: Scope,
+                    sep: String,
+                    showProject: Reference => String,
+                    mask: ScopeMask): String =
+    displayMasked(scope, sep, showProject, mask, false)
+
+  /**
+   * unified / style introduced in sbt 0.13.16 / sbt 1.0
```
Comment (dwijnand) [inline main-settings/src/main/scala/sbt/Scope.scala]: This wasn't introduced in 0.13.16

File: main-settings/src/main/scala/sbt/Scope.scala [hunk 10]
```diff
@@ -123,29 +123,84 @@ object Scope {
...
+   * By defeault, sbt will no longer display the Zero-config,
```
Comment (dwijnand) [inline main-settings/src/main/scala/sbt/Scope.scala]: default

File: main/src/main/scala/sbt/Defaults.scala [hunk 11]
```diff
@@ -2336,13 +2336,20 @@ object Classpaths {
       else Def.task((evictionWarningOptions in update).value)
     }.value
 
+    val extracted = (Project extract state0)
+    val isPlugin = sbtPlugin.value
+    val thisRef = thisProjectRef.value
+    val label =
+      if (isPlugin) Reference.display(thisRef)
+      else Def.displayRelativeReference(extracted.currentRef, thisRef)
```
Comment (dwijnand) [inline main/src/main/scala/sbt/Defaults.scala:2344]: huh, TIL that thisProjectRef != "the current project"

File: main/src/main/scala/sbt/SlashSyntax.scala [hunk 12]
```diff
@@ -0,0 +1,148 @@
+package sbt
+
+import java.io.File
+import sbt.librarymanagement.Configuration
+
+/**
+ * ScopeSyntax implements the slash syntax to scope keys for build.sbt DSL.
```
Comment (dwijnand) [inline main/src/main/scala/sbt/SlashSyntax.scala]: What is "ScopeSyntax"? Either it should read "SlashSyntax" or it needs more explanation.

File: main/src/main/scala/sbt/internal/KeyIndex.scala [hunk 13]
```diff
@@ -87,13 +109,34 @@ private[sbt] final class AKeyIndex(val data: Relation[Option[AttributeKey[_]], S
   def tasks: Set[AttributeKey[_]] = data._1s.flatten.toSet
   def tasks(key: String): Set[AttributeKey[_]] = data.reverse(key).flatten.toSet
 }
-private[sbt] final class ConfigIndex(val data: Map[Option[String], AKeyIndex]) {
+
+/*
+ * data contains the mapping between a configuration and keys.
+ * identData contains the mapping between a configuration and its identifier.
```
Comment (dwijnand) [inline main/src/main/scala/sbt/internal/KeyIndex.scala:115]: why is this necessary?

File: main/src/test/scala/ParseKey.scala [hunk 14]
```diff
@@ -22,13 +22,22 @@ object ParseKey extends Properties("Key parser test") {
...
-      import skm.{ structure, key, mask }
-
+      import skm.{ structure, key, mask => mask0 }
+
+      val hasZeroConfig = key.scope.config == Zero
+      val mask =
+        mask0 match {
+          case _ if hasZeroConfig => mask0.copy(project = true)
+          case _                  => mask0
+        }
```
Comment (dwijnand) [inline main/src/test/scala/ParseKey.scala]: this is overly complex..

```scala
val mask = if (hasZeroConfig) mask0.copy(project = true) else mask0
```

File: main/src/test/scala/ParseKey.scala [hunk 15]
```diff
@@ -22,13 +22,22 @@ object ParseKey extends Properties("Key parser test") {
...
       val expected = resolve(structure, key, mask)
-      val string = displayMasked(key, mask)
-
-      ("Key: " + displayFull(key)) |:
-        parseExpected(structure, string, expected, mask)
+      // Note that this explicitly displays the configuration axis set to Zero.
+      // This is to disambiguate `proj/Zero/name`, which could render potentially
+      // as `Zero/name`, but could be interpretted as `Zero/Zero/name`.
+      val s = displayMasked(key, mask, hasZeroConfig)
+      // println(s"s: $s")
```
Comment (dwijnand) [inline main/src/test/scala/ParseKey.scala]: remove

File: main/src/test/scala/ParseKey.scala [hunk 16]
```diff
@@ -101,7 +116,10 @@ object ParseKey extends Properties("Key parser test") {
       ("Mask: " + mask) |:
       parse(structure, s) {
       case Left(err) => false
-      case Right(sk) => Project.equal(sk, expected, mask)
+      case Right(sk) =>
+        // println(s"${sk}.key == ${expected}.key: ${sk.key == expected.key}")
+        // println(s"${sk.scope} == ${expected.scope}: ${Scope.equal(sk.scope, expected.scope, mask)}")
```
Comment (dwijnand) [inline main/src/test/scala/ParseKey.scala]: remove

File: main-settings/src/main/scala/sbt/Reference.scala (hunk 7 above)
Comment (eed3si9n) [reply main-settings/src/main/scala/sbt/Reference.scala]: Traditionally `toString` doesn't put whitespaces, so I stuck to the convention:

```
sbt:sbtRoot> eval ProjectRef(uri("uri"),"id")
[info] ans: sbt.ProjectRef = ProjectRef(uri,id)
```

File: main-settings/src/main/scala/sbt/Def.scala (hunk 8 above)
Comment (eed3si9n) [reply main-settings/src/main/scala/sbt/Def.scala]: `multi = true` is passed in whenever you have multi-projects. Prior to this change, when sbt detected a multi-project build it will display `$subproject/key` even though the method is called `displayRelative`. At least the Build URI is omitted, but I am making it *more* relative by omitting the subproject id.

Here's how this would show up. When you run `inspect test` in sbt 1.0.0 you get:

```
[info] Dependencies:
[info] 	sbtRoot/test:executeTests
[info] 	sbtRoot/test:test::streams
[info] 	sbtRoot/test:state
[info] 	sbtRoot/test:test::testResultLogger
```

In sbt 1.1.0 you'd get:

```
[info] Dependencies:
[info] 	Test/executeTests
[info] 	Test/test/streams
[info] 	Test/state
[info] 	Test/test/testResultLogger
```

I think this is better, because what's something the user can type in or use in `build.sbt`.

File: main/src/main/scala/sbt/SlashSyntax.scala (hunk 12 above)
Comment (eed3si9n) [reply main/src/main/scala/sbt/SlashSyntax.scala]: ScopeSyntax was the older name. I'll change this.

File: main/src/main/scala/sbt/internal/KeyIndex.scala (hunk 13 above)
Comment (eed3si9n) [reply main/src/main/scala/sbt/internal/KeyIndex.scala:115]: We can't change the type of `Def.ScopeKey[_]`, which binds to `sbt.Scope`, which includes `sbt.ConfigKey`, a wrapper around a `String` that represents the `name` of configuration. The mapping information between the Scala identifier and `String` thus needs to be available during conversion back and forth. For example, if the user types in `IntegrationTest/test`, the new parser needs to parse it as `ConfigKey("it")` and `AttributeKey[Unit]("test")`.

File: main/src/test/scala/ParseKey.scala (hunk 16 above)
Comment (eed3si9n) [reply main/src/test/scala/ParseKey.scala]: Actually the right thing to do is convert it into ScalaCheck label so it'll be displayed when it fails:

```scala
        (s"${sk}.key == ${expected}.key: ${sk.key == expected.key}") |:
          (s"${sk.scope} == ${expected.scope}: ${Scope.equal(sk.scope, expected.scope, mask)}") |:
          Project.equal(sk, expected, mask)
```

**Comment by typesafe-tools** [comment]
To validate this pull request against other sbt modules, please visit <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild/?sbt=pull/3434/head&deploy_to_bintray=false&run_sbt_community=false">this link.</A>

File: main-settings/src/main/scala/sbt/Reference.scala (hunk 7 above)
Comment (dwijnand) [reply main-settings/src/main/scala/sbt/Reference.scala]: It's not universally consistent:

```scala
scala> List(1, 2, 3)
res0: List[Int] = List(1, 2, 3)
```

and I can't find a single instance of creating `ProjectRef` where we _don't_ include a space: https://github.com/sbt/sbt/search?utf8=%E2%9C%93&q=ProjectRef.

Is the point not that we print what people would write?

File: main-settings/src/main/scala/sbt/Def.scala (hunk 8 above)
Comment (dwijnand) [reply main-settings/src/main/scala/sbt/Def.scala]: I see. So can we lose the commented out code and the `multi` parameter from this new `private[sbt]` method?

File: main-settings/src/main/scala/sbt/Scope.scala [hunk 17]
```diff
@@ -123,29 +123,84 @@ object Scope {
...
-      mask: ScopeMask
-  ): String = {
+  def displayMasked(scope: Scope,
+                    sep: String,
+                    showProject: Reference => String,
+                    mask: ScopeMask): String =
+    displayMasked(scope, sep, showProject, mask, false)
+
+  /**
+   * unified slash style introduced in sbt sbt 1.1.0
```
Comment (dwijnand) [inline main-settings/src/main/scala/sbt/Scope.scala]: sbt sbt

**Review by dwijnand — CHANGES_REQUESTED** [review]
2 requests, 1 more typo

File: main-settings/src/main/scala/sbt/Reference.scala (hunk 7 above)
Comment (eed3si9n) [reply main-settings/src/main/scala/sbt/Reference.scala]: I don't expect the build user to put the whitespaces around in the shell. The same goes for `/` by the way:

```
sbt:helloworld> inspect test
[info] Task: Unit
[info] Description:
[info] 	Executes all tests.
[info] Provided by:
[info] 	ProjectRef(uri("file:/Users/eed3si9n/work/hellotest/"),"root")/Test/test
[info] Defined at:
[info] 	(sbt.Defaults.testTasks) Defaults.scala:670
[info] Dependencies:
[info] 	Test/executeTests
[info] 	Test/test/streams
[info] 	Test/state
[info] 	Test/test/testResultLogger
```

Comment (eed3si9n) [reply main-settings/src/main/scala/sbt/Reference.scala]: You know what? Let me just put whitespaces to both in a separate commit. If they copy-paste to shell, shell can parse them anyway.

**Comment by typesafe-tools** [comment]
To validate this pull request against other sbt modules, please visit <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild/?sbt=pull/3434/head&deploy_to_bintray=false&run_sbt_community=false">this link.</A>

File: main-settings/src/main/scala/sbt/Reference.scala (hunk 7 above)
Comment (eed3si9n) [reply main-settings/src/main/scala/sbt/Reference.scala]: c67fe067cf63a1f18d61aaec84442348a4b09cd3

**Comment by typesafe-tools** [comment]
To validate this pull request against other sbt modules, please visit <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild/?sbt=pull/3434/head&deploy_to_bintray=false&run_sbt_community=false">this link.</A>

**Review by dwijnand — APPROVED** [review]
